SalesforceFoundation/HEDAP
# Fixed TDTM_Global_API setTdtmConfig method. 1. setTdtmConfig will clear the cached Tdtm tokens 2. setTdtmConfig will save Filter Field and Filter Value

# Critical Changes

# Changes
Added a new TDTM class, THAN_ClearCache_TDTM, that clears cached TDTM tokens after Trigger Handler record insert. Enhanced the setTdtmConfig method to clear cached TDTM tokens and to save values for Filter Field and Filter Value. Verify that your Apex unit tests and trigger handlers work as expected with these changes to setTdtmConfig.

# Issues Closed
Fixes #467
Fixes #479

# New Metadata
THAN_ClearCache_TDTM
THAN_ClearCache_TEST

# Deleted Metadata

# Testing Notes
Please note that this PR includes the fix for two GitHub issues (#467 and #479).  Those two issues are related to the TDTM_Global_API that will be used in Apex unit test. Those two issues are not directly related to the Filter logic in TDTM framework. 

#467 : The setTDTMConfig method missed the Filter Field and Filter Value when save the tokens into the database.

Steps to reproduce:
1. Run TDTM_Global_API.setTdtmConfig in the anonymous apex window (or in the Apex unit test) with a configuration that includes a value for the Filter Field and Filter Value fields in some of the items.

```
List<TDTM_Global_API.TdtmToken> tokens = new List<TDTM_Global_API.TdtmToken>();
tokens.add(new TDTM_Global_API.TdtmToken('Filter_Test_TDTM', 'Contact',  'BeforeInsert', 1, true, false, 'LastName', 'Testerson', false,  ''));
TDTM_Global_API.setTdtmConfig(tokens);
```

2. Do a query about the Trigger_handler__c record **or look at the Trigger Handlers tab**. You will find that the Filter Field and Filter Value is not in the database. 
3. The root cause is that TDTM_Manager.mergeToken method excluding those fields.

#479 :  The getTDTMConfig always return the cached TDTM tokens.

Steps to reproduce:
1. Insert a Trigger_Handler__c record (for example update a field on Contact when insert Contact in Apex unit test). The insertion operation will cause system to populate and cache the TDTM Tokens with default config. 
2. Insert a Contact record **in the same transaction**. You will find that the new Trigger Handler didn’t run. This is because the system uses the cached TDTM tokens from step 1 and the cache does not include the new Trigger_Hanlder__c record. 
3. When you call ```TDTM_Global_API.getTdtmConfig``` method, it will always return the cached token from step 1 unless you manipulate the cached token directly. 

Another way to reproduce **- do all these steps in the same transaction (anonymous Apex or unit test)**:
1. Insert a Contact record or call ```TDTM_Global_API.getTDTMConfig```. The operation will cause system to populate and cache the TDTM Tokens with default config. 
2. Build a TDTM Token
3. Set the new Token
4. When you call ```TDTM_Global_API.getTDTMConfig```, you will get the cached token from step 1 instead of the new one inserted in step 2:
  
```
TDTM_Global_API.getDefaultTdtmConfig();
List<TDTM_Global_API.TdtmToken> tokens = new List<TDTM_Global_API.TdtmToken>();
tokens.add(new TDTM_Global_API.TdtmToken('Filter_Test_TDTM', 'Contact', 
'BeforeInsert', 1, true, false, 'LastName', 'Testerson', false,  ''));
TDTM_Global_API.setTdtmConfig(tokens);
List<TDTM_Global_API.TdtmToken> config = TDTM_Global_API.getTDTMConfig();
System.debug('****config: ' + JSON.serializePretty(config));
```
Basically, the TDTM_Global_API setTdtmConfig is an utility method for a developer to implement their Apex unit test. The end user will not use it. We already included our unit test to test the logic inside the function. The developer can implement additional unit test to verify the fixes in this PR if needed. 

For QA, we need to **additionally** verify four things for this PR:

1. Verify the existing regression test cases still pass
2. Verify the existing default Trigger Handlers stay the same after HEDA upgrade
3. Verify that a New TDTM Trigger Handler (THAN_ClearCache_TDTM) is added after the HEDA upgrade
4. If user created their own Trigger Handlers, verify they will stay the same after HEDA upgrade
